File: rb/RailwayStationPhotos_sdk.rb
# RailwayStationPhotos SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'RailwayStationPhotos_types'


class RailwayStationPhotosSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = RailwayStationPhotosUtility.new
    @_utility = utility

    config = RailwayStationPhotosConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features in the resolved order (make_options puts an explicit array
    # order first, else defaults to test-first). Ordering matters: the `test`
    # feature installs the base mock transport and the transport features
    # (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    # must be added before them to sit at the base of the chain.
    feature_opts = RailwayStationPhotosHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      featureorder = VoxgigStruct.getpath(@options, "__derived__.featureorder")
      if featureorder.is_a?(Array)
        featureorder.each do |fname|
          fopts = RailwayStationPhotosHelpers.to_map(feature_opts[fname])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, RailwayStationPhotosFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    RailwayStationPhotosUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = RailwayStationPhotosHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = RailwayStationPhotosHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = RailwayStationPhotosHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = RailwayStationPhotosSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue RailwayStationPhotosError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = RailwayStationPhotosHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = RailwayStationPhotosHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.AdminInbox.list / client.AdminInbox.load({ "id" => ... })
  def AdminInbox(data = nil)
    require_relative 'entity/admin_inbox_entity'
    AdminInboxEntity.new(self, data)
  end


  # Canonical facade: client.Country.list / client.Country.load({ "id" => ... })
  def Country(data = nil)
    require_relative 'entity/country_entity'
    CountryEntity.new(self, data)
  end


  # Canonical facade: client.Inbox.list / client.Inbox.load({ "id" => ... })
  def Inbox(data = nil)
    require_relative 'entity/inbox_entity'
    InboxEntity.new(self, data)
  end


  # Canonical facade: client.InboxCount.list / client.InboxCount.load({ "id" => ... })
  def InboxCount(data = nil)
    require_relative 'entity/inbox_count_entity'
    InboxCountEntity.new(self, data)
  end


  # Canonical facade: client.InboxEntry.list / client.InboxEntry.load({ "id" => ... })
  def InboxEntry(data = nil)
    require_relative 'entity/inbox_entry_entity'
    InboxEntryEntity.new(self, data)
  end


  # Canonical facade: client.InboxStateQuery.list / client.InboxStateQuery.load({ "id" => ... })
  def InboxStateQuery(data = nil)
    require_relative 'entity/inbox_state_query_entity'
    InboxStateQueryEntity.new(self, data)
  end


  # Canonical facade: client.OAuthToken.list / client.OAuthToken.load({ "id" => ... })
  def OAuthToken(data = nil)
    require_relative 'entity/o_auth_token_entity'
    OAuthTokenEntity.new(self, data)
  end


  # Canonical facade: client.Oauth.list / client.Oauth.load({ "id" => ... })
  def Oauth(data = nil)
    require_relative 'entity/oauth_entity'
    OauthEntity.new(self, data)
  end


  # Canonical facade: client.Photo.list / client.Photo.load({ "id" => ... })
  def Photo(data = nil)
    require_relative 'entity/photo_entity'
    PhotoEntity.new(self, data)
  end


  # Canonical facade: client.PhotoDownload.list / client.PhotoDownload.load({ "id" => ... })
  def PhotoDownload(data = nil)
    require_relative 'entity/photo_download_entity'
    PhotoDownloadEntity.new(self, data)
  end


  # Canonical facade: client.PhotoStation.list / client.PhotoStation.load({ "id" => ... })
  def PhotoStation(data = nil)
    require_relative 'entity/photo_station_entity'
    PhotoStationEntity.new(self, data)
  end


  # Canonical facade: client.PhotoUpload.list / client.PhotoUpload.load({ "id" => ... })
  def PhotoUpload(data = nil)
    require_relative 'entity/photo_upload_entity'
    PhotoUploadEntity.new(self, data)
  end


  # Canonical facade: client.Photographer.list / client.Photographer.load({ "id" => ... })
  def Photographer(data = nil)
    require_relative 'entity/photographer_entity'
    PhotographerEntity.new(self, data)
  end


  # Canonical facade: client.Profile.list / client.Profile.load({ "id" => ... })
  def Profile(data = nil)
    require_relative 'entity/profile_entity'
    ProfileEntity.new(self, data)
  end


  # Canonical facade: client.PublicInbox.list / client.PublicInbox.load({ "id" => ... })
  def PublicInbox(data = nil)
    require_relative 'entity/public_inbox_entity'
    PublicInboxEntity.new(self, data)
  end


  # Canonical facade: client.Stat.list / client.Stat.load({ "id" => ... })
  def Stat(data = nil)
    require_relative 'entity/stat_entity'
    StatEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = RailwayStationPhotosSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: lua/railway-station-photos_sdk.lua
-- RailwayStationPhotos SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local RailwayStationPhotosSDK = {}
RailwayStationPhotosSDK.__index = RailwayStationPhotosSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

RailwayStationPhotosSDK._make_feature = _make_feature


function RailwayStationPhotosSDK.new(options)
  local self = setmetatable({}, RailwayStationPhotosSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features in the resolved order (make_options puts an explicit list
  -- order first, else defaults to test-first). Ordering matters: the `test`
  -- feature installs the base mock transport and the transport features
  -- (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
  -- must be added before them to sit at the base of the chain.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local featureorder = vs.getpath(self.options, "__derived__.featureorder")
    if type(featureorder) == "table" then
      for _, fname in ipairs(featureorder) do
        local fopts = helpers.to_map(feature_opts[fname])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function RailwayStationPhotosSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function RailwayStationPhotosSDK:get_utility()
  return Utility.copy(self._utility)
end


function RailwayStationPhotosSDK:get_root_ctx()
  return self._rootctx
end


function RailwayStationPhotosSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function RailwayStationPhotosSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:AdminInbox():list() / client:AdminInbox():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:AdminInbox(data)
  local EntityMod = require("entity.admin_inbox_entity")
  if data == nil then
    if self._admin_inbox == nil then
      self._admin_inbox = EntityMod.new(self, nil)
    end
    return self._admin_inbox
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Country():list() / client:Country():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Country(data)
  local EntityMod = require("entity.country_entity")
  if data == nil then
    if self._country == nil then
      self._country = EntityMod.new(self, nil)
    end
    return self._country
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Inbox():list() / client:Inbox():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Inbox(data)
  local EntityMod = require("entity.inbox_entity")
  if data == nil then
    if self._inbox == nil then
      self._inbox = EntityMod.new(self, nil)
    end
    return self._inbox
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:InboxCount():list() / client:InboxCount():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:InboxCount(data)
  local EntityMod = require("entity.inbox_count_entity")
  if data == nil then
    if self._inbox_count == nil then
      self._inbox_count = EntityMod.new(self, nil)
    end
    return self._inbox_count
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:InboxEntry():list() / client:InboxEntry():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:InboxEntry(data)
  local EntityMod = require("entity.inbox_entry_entity")
  if data == nil then
    if self._inbox_entry == nil then
      self._inbox_entry = EntityMod.new(self, nil)
    end
    return self._inbox_entry
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:InboxStateQuery():list() / client:InboxStateQuery():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:InboxStateQuery(data)
  local EntityMod = require("entity.inbox_state_query_entity")
  if data == nil then
    if self._inbox_state_query == nil then
      self._inbox_state_query = EntityMod.new(self, nil)
    end
    return self._inbox_state_query
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:OAuthToken():list() / client:OAuthToken():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:OAuthToken(data)
  local EntityMod = require("entity.o_auth_token_entity")
  if data == nil then
    if self._o_auth_token == nil then
      self._o_auth_token = EntityMod.new(self, nil)
    end
    return self._o_auth_token
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Oauth():list() / client:Oauth():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Oauth(data)
  local EntityMod = require("entity.oauth_entity")
  if data == nil then
    if self._oauth == nil then
      self._oauth = EntityMod.new(self, nil)
    end
    return self._oauth
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Photo():list() / client:Photo():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Photo(data)
  local EntityMod = require("entity.photo_entity")
  if data == nil then
    if self._photo == nil then
      self._photo = EntityMod.new(self, nil)
    end
    return self._photo
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PhotoDownload():list() / client:PhotoDownload():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:PhotoDownload(data)
  local EntityMod = require("entity.photo_download_entity")
  if data == nil then
    if self._photo_download == nil then
      self._photo_download = EntityMod.new(self, nil)
    end
    return self._photo_download
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PhotoStation():list() / client:PhotoStation():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:PhotoStation(data)
  local EntityMod = require("entity.photo_station_entity")
  if data == nil then
    if self._photo_station == nil then
      self._photo_station = EntityMod.new(self, nil)
    end
    return self._photo_station
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PhotoUpload():list() / client:PhotoUpload():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:PhotoUpload(data)
  local EntityMod = require("entity.photo_upload_entity")
  if data == nil then
    if self._photo_upload == nil then
      self._photo_upload = EntityMod.new(self, nil)
    end
    return self._photo_upload
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Photographer():list() / client:Photographer():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Photographer(data)
  local EntityMod = require("entity.photographer_entity")
  if data == nil then
    if self._photographer == nil then
      self._photographer = EntityMod.new(self, nil)
    end
    return self._photographer
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Profile():list() / client:Profile():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Profile(data)
  local EntityMod = require("entity.profile_entity")
  if data == nil then
    if self._profile == nil then
      self._profile = EntityMod.new(self, nil)
    end
    return self._profile
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:PublicInbox():list() / client:PublicInbox():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:PublicInbox(data)
  local EntityMod = require("entity.public_inbox_entity")
  if data == nil then
    if self._public_inbox == nil then
      self._public_inbox = EntityMod.new(self, nil)
    end
    return self._public_inbox
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Stat():list() / client:Stat():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function RailwayStationPhotosSDK:Stat(data)
  local EntityMod = require("entity.stat_entity")
  if data == nil then
    if self._stat == nil then
      self._stat = EntityMod.new(self, nil)
    end
    return self._stat
  end
  return EntityMod.new(self, data)
end




function RailwayStationPhotosSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = RailwayStationPhotosSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return RailwayStationPhotosSDK
